File: Smallest distinct window - GFG/smallest-distinct-window.py
#User function Template for python3

class Solution:
    def findSubString(self, str):
        # Your code goes here
        target=len(set(str))
        n=len(str)
        start=0
        end=0
        visited=dict()
        mini=float('inf')
        count=0
        while end<n:
            visited[str[end]]=visited.get(str[end],0)+1
            while visited[str[start]]>1:
                visited[str[start]]-=1
                start+=1
            if len(visited)==target:
                mini=min(mini,end-start+1)
            end+=1    
                    
                
            
        return mini    
                
            
            
    
    
    

#{ 
#  Driver Code Starts
#Initial Template for Python 3


def main():

    T = int(input())

    while(T > 0):
    	str = input()
    	ob=Solution()
    	print(ob.findSubString(str))
    	
    	T -= 1


if __name__ == "__main__":
    main()

# } Driver Code Ends

File: Smallest distinct window - GFG/smallest-distinct-window.cpp
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
class Solution{
    public:
    int findSubString(string str)
    {
        // Your code goes here   
        unordered_map<char,int> mp;
        for(int i=0;i<str.size();i++){
            if (mp.find(str[i])==mp.end()){
                mp.insert({str[i],1});
                
            }
        }
        int n=mp.size();
        unordered_map<char,int> new_map;
        int end=0;
        int start=0;
        int ans=INT_MAX;
        while (end<str.size()){
            if(mp.find(str[end])!=mp.end()){
                new_map[str[end]]++;
            }
            
            while (new_map[str[start]]>1){
                new_map[str[start]]--;
                start++;
            }
            if (new_map.size()==n){
                ans=min(ans,end-start+1);
            }
            end++;
        }
        return ans;
        
    }
};

// { Driver Code Starts.
// Driver code
int main() {
    int t;
    cin >> t;
    while (t--) {

        string str;
        cin >> str;
        Solution ob;
        cout << ob.findSubString(str) << endl;
    }
    return 0;
}  // } Driver Code Ends